### ISSUE TITLE
Panic when hash join is performed on materialized view with group by clause

### ISSUE THREAD

**Issue by Graphcalibur** [singularity-data/risingwave#4213]
**Describe the bug**
Hash joining a table with a materialized view with a `group by` clause causes a panic

**To Reproduce**
```
create table t1 (v1 int);
create table t2 (v1 int);
create materialized view t3 as select v1 from t2 group by v1;
select * from t1 join t3 on t1.v1 = t3.v1;
```

**Expected behavior**
Query runs successfully.

**Comment by skyzh** [comment]
Do you have any logs / information about the error?

**Comment by Graphcalibur** [comment]
This was the stack backtrace in the playground:

```
thread 'tokio-runtime-worker' panicked at 'index out of bounds: the len is 1 but the index is 3', src/batch/src/task/task_execution.rs:363:24
stack backtrace:
   0: rust_begin_unwind
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/core/src/panicking.rs:142:14
   2: core::panicking::panic_bounds_check
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/core/src/panicking.rs:84:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index_mut
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/core/src/slice/index.rs:248:14
   4: core::slice::index::<impl core::ops::index::IndexMut<I> for [T]>::index_mut
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/alloc/src/vec/mod.rs:2639:9
   5: <alloc::vec::Vec<T,A> as core::ops::index::IndexMut<I>>::index_mut
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/alloc/src/vec/mod.rs:2639:9
   6: risingwave_batch::task::task_execution::BatchTaskExecution<C>::get_task_output
             at ./src/batch/src/task/task_execution.rs:363:24
   7: risingwave_batch::task::task_manager::BatchManager::take_output
             at ./src/batch/src/task/task_manager.rs:99:9
   8: <risingwave_batch::task::context::ComputeNodeContext as risingwave_batch::task::context::BatchTaskContext>::get_task_output
             at ./src/batch/src/task/context.rs:75:9
   9: risingwave_batch::execution::local_exchange::LocalExchangeSource::create
             at ./src/batch/src/execution/local_exchange.rs:38:27
  10: <risingwave_batch::executor::generic_exchange::DefaultCreateSource as risingwave_batch::executor::generic_exchange::CreateSource>::create_source::{{closure}}
             at ./src/batch/src/executor/generic_exchange.rs:72:42
  11: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/core/src/future/mod.rs:91:19
  12: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/core/src/future/future.rs:124:9
  13: risingwave_batch::executor::generic_exchange::GenericExchangeExecutor<CS,C>::do_execute::{{closure}}
             at ./src/batch/src/executor/generic_exchange.rs:148:18
  14: <futures_async_stream::try_stream::from_generator::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-async-stream-0.2.5/src/lib.rs:430:31
  15: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.21/src/stream.rs:120:9
  16: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/mod.rs:1626:9
  17: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/next.rs:32:9
  18: <tracing_futures::Instrumented<T> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-futures-0.2.5/src/lib.rs:283:9
  19: risingwave_batch::executor::trace::TraceExecutor::do_execute::{{closure}}
             at ./src/batch/src/executor/trace.rs:66:14
  20: <futures_async_stream::try_stream::from_generator::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-async-stream-0.2.5/src/lib.rs:430:31
  21: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.21/src/stream.rs:120:9
  22: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/mod.rs:1626:9
  23: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/next.rs:32:9
  24: risingwave_batch::executor::join::hash_join::HashJoinExecutor<K>::do_execute::{{closure}}
             at ./src/batch/src/executor/join/hash_join.rs:178:40
  25: <futures_async_stream::try_stream::from_generator::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-async-stream-0.2.5/src/lib.rs:430:31
  26: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.21/src/stream.rs:120:9
  27: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/mod.rs:1626:9
  28: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/next.rs:32:9
  29: <tracing_futures::Instrumented<T> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-futures-0.2.5/src/lib.rs:283:9
  30: risingwave_batch::executor::trace::TraceExecutor::do_execute::{{closure}}
             at ./src/batch/src/executor/trace.rs:66:14
  31: <futures_async_stream::try_stream::from_generator::GenTryStream<G> as futures_core::stream::Stream>::poll_next
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-async-stream-0.2.5/src/lib.rs:430:31
  32: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.21/src/stream.rs:120:9
  33: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/mod.rs:1626:9
  34: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/next.rs:32:9
  35: risingwave_batch::task::task_execution::BatchTaskExecution<C>::try_execute::{{closure}}::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/macros/select.rs:507:49
  36: <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/future/poll_fn.rs:38:9
  37: risingwave_batch::task::task_execution::BatchTaskExecution<C>::try_execute::{{closure}}
             at ./src/batch/src/task/task_execution.rs:292:13
  38: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/core/src/future/mod.rs:91:19
  39: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/task/task_local.rs:345:35
  40: tokio::task::task_local::LocalKey<T>::scope_inner
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/task/task_local.rs:231:19
  41: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/task/task_local.rs:342:13
  42: <tracing_futures::Instrumented<T> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-futures-0.2.5/src/lib.rs:283:9
  43: risingwave_batch::task::task_execution::BatchTaskExecution<C>::async_execute::{{closure}}::{{closure}}::{{closure}}
             at ./src/batch/src/task/task_execution.rs:267:23
  44: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/core/src/future/mod.rs:91:19
  45: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/core/src/future/future.rs:124:9
  46: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.34/src/instrument.rs:272:9
  47: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/core.rs:165:17
  48: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/loom/std/unsafe_cell.rs:14:9
  49: tokio::runtime::task::core::CoreStage<T>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/core.rs:155:13
  50: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/harness.rs:480:19
  51: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/core/src/panic/unwind_safe.rs:271:9
  52: std::panicking::try::do_call
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/std/src/panicking.rs:492:40
  53: __rust_try
  54: std::panicking::try
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/std/src/panicking.rs:456:19
  55: std::panic::catch_unwind
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/std/src/panic.rs:137:14
  56: tokio::runtime::task::harness::poll_future
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/harness.rs:468:18
  57: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/harness.rs:104:27
  58: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/harness.rs:57:15
  59: tokio::runtime::task::raw::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/raw.rs:144:5
  60: tokio::runtime::task::raw::RawTask::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/raw.rs:84:18
  61: tokio::runtime::task::LocalNotified<S>::run
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/mod.rs:381:9
  62: tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/thread_pool/worker.rs:458:21
  63: tokio::coop::with_budget::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/coop.rs:102:9
  64: std::thread::local::LocalKey<T>::try_with
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/std/src/thread/local.rs:445:16
  65: std::thread::local::LocalKey<T>::with
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/std/src/thread/local.rs:421:9
  66: tokio::coop::with_budget
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/coop.rs:95:5
  67: tokio::coop::budget
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/coop.rs:72:5
  68: tokio::runtime::thread_pool::worker::Context::run_task
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/thread_pool/worker.rs:434:9
  69: tokio::runtime::thread_pool::worker::Context::run
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/thread_pool/worker.rs:408:24
  70: tokio::runtime::thread_pool::worker::run::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/thread_pool/worker.rs:386:17
  71: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/macros/scoped_tls.rs:61:9
  72: tokio::runtime::thread_pool::worker::run
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/thread_pool/worker.rs:383:5
  73: tokio::runtime::thread_pool::worker::Launch::launch::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/thread_pool/worker.rs:362:45
  74: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/blocking/task.rs:42:21
  75: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.34/src/instrument.rs:272:9
  76: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/core.rs:165:17
  77: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/loom/std/unsafe_cell.rs:14:9
  78: tokio::runtime::task::core::CoreStage<T>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/core.rs:155:13
  79: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/harness.rs:480:19
  80: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/core/src/panic/unwind_safe.rs:271:9
  81: std::panicking::try::do_call
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/std/src/panicking.rs:492:40
  82: __rust_try
  83: std::panicking::try
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/std/src/panicking.rs:456:19
  84: std::panic::catch_unwind
             at /rustc/263edd43c5255084292329423c61a9d69715ebfa/library/std/src/panic.rs:137:14
  85: tokio::runtime::task::harness::poll_future
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/harness.rs:468:18
  86: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/harness.rs:104:27
  87: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/harness.rs:57:15
  88: tokio::runtime::task::raw::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/raw.rs:144:5
  89: tokio::runtime::task::raw::RawTask::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/raw.rs:84:18
  90: tokio::runtime::task::UnownedTask<S>::run
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/task/mod.rs:418:9
  91: tokio::runtime::blocking::pool::Task::run
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/blocking/pool.rs:91:9
  92: tokio::runtime::blocking::pool::Inner::run
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/blocking/pool.rs:325:17
  93: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.0/src/runtime/blocking/pool.rs:300:13
```

**Comment by skyzh** [comment]
Maybe the schema passed to exchange is wrong? cc @liurenjie1024 @wzzzzd

**Comment by liurenjie1024** [comment]
@wzzzzd  PTAL when refactoring hash join.

**Comment by st1page** [comment]
the query's plan is 
```
dev=> explain select * from t1 join t3 on t1.v1 = t3.v1;
                        QUERY PLAN                         
-----------------------------------------------------------
 BatchExchange { order: [], dist: Single }
   BatchHashJoin { type: Inner, predicate: t1.v1 = t3.v1 }
     BatchExchange { order: [], dist: HashShard(t1.v1) }
       BatchScan { table: t1, columns: [v1] }
     BatchScan { table: t3, columns: [v1] }
```
With the same distribution, the  `BatchHashJoin` and its right side's `BatchScan` and in the same fragment.
So I force to split the hash-join and scan into 2 fragments by some tricky method, and the error disappears. 
```
dev=> explain select * from t1 join t3 on t1.v1 = t3.v1;
                        QUERY PLAN                         
-----------------------------------------------------------
 BatchExchange { order: [], dist: Single }
   BatchHashJoin { type: Inner, predicate: t1.v1 = t3.v1 }
     BatchExchange { order: [], dist: HashShard(t1.v1) }
       BatchScan { table: t1, columns: [v1] }
     BatchExchange { order: [], dist: HashShard(t3.v1) }
       BatchScan { table: t3, columns: [v1] }
```
the hash join's two sides should have the same parallelism and sharding rule
I'm not familiar with the scheduler part in our system. do you have any ideas? @xxchan @liurenjie1024

**Comment by xxchan** [comment]
This happends because:
1. When building the stage containing the HashJoin, the parallism is initially 1.
2. When visiting the Exchange node connecting the stage of Scan(t1), we have 
```rust 
let child_exchange_info = node.distribution().to_prost(builder.parallelism);
```
So the tasks will have 1 outputs.

3. When visiting the other side containing the other scan node, we updated `builder.table_scan_info`
4. When finish building, **we modified the parallelism of the stage**
```rust
parallelism: match &self.table_scan_info {
    None => self.parallelism,
    Some(info) => info
        .partitions
        .as_ref()
        .map(|m| m.len() as u32)
        .unwrap_or(1),
},
```
So we have 8 (the number of partitions of the table) tasks for the fragment containing the HashJoin, but the exchange will only have 1 output, and then it panics.

This can explain why it's OK if both **or neither** sides of the HashJoin have Exchange.
